### PR TITLE
Poll Pipelined from Sessiond

### PR DESCRIPTION
**Pull request by veshkemburu**
Signed-off-by: veshkemburu <veshkemburu@fb.com>

## Summary

Pull model code to request stats in sessiond from pipelined. Needs to connect to grpc endpoint created(https://github.com/magma/magma/pull/7572)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 1]
```diff
@@ -147,6 +147,44 @@ void LocalEnforcer::setup(
...
+      return;
+    }
+    auto session_map = session_store_.read_all_sessions();
+    SessionUpdate update =
+        SessionStore::get_default_session_update(session_map);
+    MLOG(MDEBUG) << "Aggregating " << request_cpy.records_size() << " records";
+    enforcer_->aggregate_records(session_map, request_cpy, update);
+    check_usage_for_reporting(std::move(session_map), update);
+  });
+}
```
Comment (uri200) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp:157]: Will this function eventually replace void LocalSessionManagerHandlerImpl::ReportRuleStats?

File: lte/gateway/c/session_manager/sessiond_main.cpp [hunk 2]
```diff
@@ -229,6 +229,18 @@ int main(int argc, char* argv[]) {
...
+  auto rule_store = std::make_shared<magma::StaticRuleStore>();
+  magma::PollStats periodic_stats_requester;
+  std::thread periodic_stats_requester_thread([&]() {
+    periodic_stats_requester.start_loop(
+      [&](std::vector<magma::PollStats> request){
+        //call enforcer function
+      },
+      config["rule_update_inteval_sec"].as<uint32_t>());
+      periodic_stats_requester.stop();
+  });
```
Comment (uri200) [inline lte/gateway/c/session_manager/sessiond_main.cpp]: why is this on the main. Wouldn't it be better if this were ran from localEnforcer? This way in the future we may be able to control the poll interval dynamically.

Or Maybe you can create a `run` method on  to start this on localEnforcer

File: lte/gateway/c/session_manager/sessiond_main.cpp [hunk 3]
```diff
@@ -229,6 +229,18 @@ int main(int argc, char* argv[]) {
     policy_loader.stop();
   });
 
+  //Start off a thread to periodically poll stats from Pipelined
+  auto rule_store = std::make_shared<magma::StaticRuleStore>();
```
Comment (uri200) [inline lte/gateway/c/session_manager/sessiond_main.cpp]: why is this needed?

File: lte/gateway/c/session_manager/PollStatsThread.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/**
```
Comment (koolzz) [inline lte/gateway/c/session_manager/PollStatsThread.h]: I think PollStats.h would be better name

File: lte/gateway/c/session_manager/LocalEnforcer.h [hunk 5]
```diff
@@ -294,6 +294,7 @@ class LocalEnforcer {
       SessionMap& session_map, const PolicyBearerBindingRequest& request,
       SessionUpdate& session_update);
 
+  void PollStats();
```
Comment (koolzz) [inline lte/gateway/c/session_manager/LocalEnforcer.h]: don't use camel case for naming functions

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 6]
```diff
@@ -147,6 +147,10 @@ void LocalEnforcer::setup(
   }
 }
 
+void LocalEnforcer::PollStats() {
+  pipelined_client_->poll_stats(0, 0, NULL);
```
Comment (koolzz) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp]: add some type of a default callback it can be simple for now
```
 pipelined_client_->poll_stats(0, 0, [](Status status, RuleRecordTable resp) {
        if (!status.ok()) {
          MLOG(MERROR) << "Coudn't poll stats " << ": " << status.error_message();
        }
      });
```

File: lte/gateway/c/session_manager/PollStats.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <stdint.h>                      // for uint32_t
```
Comment (koolzz) [inline lte/gateway/c/session_manager/PollStats.cpp]: Remove the for uin32_t and alike comments

File: lte/gateway/c/session_manager/PollStats.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <stdint.h>                      // for uint32_t
+#include <atomic>                        // for atomic
```
Comment (koolzz) [inline lte/gateway/c/session_manager/PollStats.h]: Same remove the //for comments

File: lte/gateway/c/session_manager/sessiond_main.cpp [hunk 9]
```diff
@@ -321,6 +322,14 @@ int main(int argc, char* argv[]) {
     }
   });
 
+  // Start off a thread to periodically poll stats from Pipelined
+  // every fixed interval of time
+  auto periodic_stats_requester = std::make_shared<magma::PollStats>();
+  std::thread periodic_stats_requester_thread([&]() {
+    periodic_stats_requester->start_loop(local_enforcer, 5);
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/sessiond_main.cpp:340]: How did we decide on this 5 value? Is this the default polling config? Will we be loading this from a config eventually?

File: lte/gateway/c/session_manager/PollStats.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <stdint.h>
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/PollStats.h]: Do we need all the includes here?

**Review by themarwhal — CHANGES_REQUESTED** [review]
Change overall looks good!

I would rename `PollStats` to `StatsPoller`

File: lte/gateway/c/session_manager/sessiond_main.cpp (hunk 9 above)
Comment (veshkemburu) [reply lte/gateway/c/session_manager/sessiond_main.cpp:340]: Yes we will load from a config eventually, I just used a random value for now

File: lte/gateway/c/session_manager/StatsPoller.h [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+class StatsPoller {
+ public:
+  /**
+   * start_loop is the main function to call to initiate a loop. Based
+   * on the given loop interval length, this function will poll stats from
+   * Pipelined every loop_interval_seconds
+   */
+  void start_loop(
+      std::shared_ptr<LocalEnforcer> local_enforcer,
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/StatsPoller.h:27]: Can you address the cpplint here?
` Add #include <memory> for shared_ptr<>  [build/include_what_you_use] [4]`

File: lte/gateway/c/session_manager/sessiond_main.cpp (hunk 9 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/sessiond_main.cpp:340]: Can you add a comment to indicate this?

**Review by themarwhal — CHANGES_REQUESTED** [review]
2 more comments, then it is good to go :)

**Review by themarwhal — APPROVED** [review]
